src/query/storages/fuse/src/operations/read/parquet_data_source_deserializer.rs

```diff
@@ -62,6 +72,46 @@ pub struct DeserializeDataTransform {
     virtual_reader: Arc<Option<VirtualColumnReader>>,
 
     base_block_ids: Option<Scalar>,
+    cached_runtime_filter: Option<Expr>,
+}
+
+#[macro_export]
+macro_rules! impl_runtime_filter {
+    () => {
+        fn runtime_filter(&mut self, data_block: DataBlock) -> Result<DataBlock> {
+            // Check if already cached runtime filters
+            if self.cached_runtime_filter.is_none() {
+                let runtime_filters = self.ctx.get_runtime_filter_with_id(self.table_index);
+                let filter = runtime_filters
+                    .iter()
+                    .map(|filter| {
+                        cast_expr_to_non_null_boolean(
+                            filter
+                                .project_column_ref(|name| self.src_schema.index_of(name).unwrap()),
+                        )
+                        .unwrap()
+                    })
+                    .try_reduce(|lhs, rhs| {
+                        check_function(None, "and_filters", &[], &[lhs, rhs], &BUILTIN_FUNCTIONS)
+                    })
+                    .transpose()
+                    .unwrap_or_else(|| {
+                        Err(ErrorCode::Internal(
+                            "Invalid empty predicate list".to_string(),
+                        ))
+                    })?;
+                self.cached_runtime_filter = Some(filter);
+            }
+            // Using runtime filter to filter data_block
+            let func_ctx = self.ctx.get_function_context()?;
+            let evaluator = Evaluator::new(&data_block, &func_ctx, &BUILTIN_FUNCTIONS);
+            let filter = evaluator
```

Comment:
We perform one more filter on all data on the probe side, will this always reduce the overall query time ?

Reply:
> We perform extra filter on all data on the probe side, will this always reduce the overall query time ?

Reply:
Update performance explanation in summary.